lib/Sema/TypeOfReference.cpp

```diff
@@ -977,6 +977,12 @@ ConstraintSystem::getTypeOfReference(ValueDecl *value,
     return { openedType, openedType, openedType, openedType, Type() };
   }
 
+  // FIXME: implement this properly
```

Comment:
@DougGregor I crashed here after rebasing onto your latest changes. The confident comment that VarDecl is the only remaining possibility makes me suspect this is not the right place to fix this however, so I didn't put much effort to figure out the Right Fix(tm).

Reply:
Okay. Do you need some help with something here, or was this just an FYI?

Reply:
If there's anything that strikes you just from looking at it, about why we'd get this error for the special attribute source files but not otherwise, that'd be useful context. If not, I'll take some more time to investigate after the break.

Reply:
I guess my point is, I suspect that the actual fix should be somewhere else, since EnumElements normally don't end up here. I just don't know where, yet.

Reply:
The issue seems to be related to the cached parsing we do, because the crash only occurs when compiling a file with 2 identical macros, and only when those macros are applied by ClangImporter (so manually spelling the macro out in the source file doesn't trigger the crash).

When we resolve the first macro, we run typeCheckExpression on this expr:
```
(macro_expansion_expr type="<null>" name="PointerBounds" discriminator=65535
  (argument_list
    (argument
      (call_expr type="<null>" isolation_crossing="none"
        (unresolved_member_expr type="<null>" name="countedBy" function_ref=unapplied)
        (argument_list labels="pointer:count:"
          (argument label="pointer"
            (integer_literal_expr type="<null>" value="2" builtin_initializer="**NULL**" initializer="**NULL**"))
          (argument label="count"
            (string_literal_expr type="<null>" encoding=utf8 value="len" builtin_initializer="**NULL**" initializer="**NULL**")))))))
```. When we resolve the second one, the AST nodes are already resolved:
```
(macro_expansion_expr type="<null>" name="PointerBounds" discriminator=65535
  (argument_list
    (argument
      (call_expr type="PointerParam" location=:1:17 range=[:1:16 - line:1:51] isolation_crossing="none"
        (dot_syntax_call_expr type="(Int, String) -> PointerParam" location=:1:17 range=[:1:16 - line:1:17] isolation_crossing="none"
          (declref_expr type="(PointerParam.Type) -> (Int, String) -> PointerParam" location=:1:17 range=[:1:17 - line:1:17] decl="Swift.(file).PointerParam.countedBy(pointer:count:)" function_ref=single)
          (argument_list implicit
            (argument
              (type_expr implicit type="PointerParam.Type" location=:1:16 range=[:1:16 - line:1:16] typerepr="PointerParam"))))
        (argument_list labels="pointer:count:"
          (argument label="pointer"
            (integer_literal_expr type="Int" location=:1:36 range=[:1:36 - line:1:36] value="2" builtin_initializer="Swift.(file).Int.init(_builtinIntegerLiteral:)" initializer="**NULL**"))
          (argument label="count"
            (string_literal_expr type="String" location=:1:46 range=[:1:46 - line:1:46] encoding=utf8 value="len" builtin_initializer="Swift.(file).String extension.init(_builtinStringLiteral:utf8CodeUnitCount:isASCII:)" initializer="**NULL**")))))))
```
This infrastructure is not set up to handle resolved AST nodes like `dot_syntax_call_expr`, because it should always run before those are created. For some reason the result of the `ResolveMacroRequest` doesn't seem to be cached properly, despite seemingly having the same inputs. Looking into why that is.

Reply:
Ah, I understand this now. It's because we aren't properly cloning the expression within each CustomAttr node. I'll think about how to address this.

Reply:
I tried doing the opposite, and not clone the CustomAttr node in the first place. At least at first glance, it seems to work and the ResolveMacroRequest is cached as it should. Would this be problematic for some reason I'm missing?

Reply:
I expect that means that you aren't getting the attribute applied?